### PR TITLE
Fixes #22847 - Use pf-react donut chart in facts page

### PR DESCRIPTION
**Pull request by lizagilman**

### CODE REVIEW

**Comment by ohadlevy** [comment]
@lizagilman thanks for the pr! @danseethaler can you please review?

**Comment by danseethaler** [comment]
hey @lizagilman - you should be able to run `npm run lint` locally to fix up these HoundCI issues before you push. I've forgotten that lots of times 😀

I think it would be helpful to add a storybook to demo the chart changes you're making here. Let me know your thoughts on that and if you need any direction. I don't see any way to demo the changes currently.

Also, I don't see the `c3` dependency actually being used in the code. Is it coming from other components that you're pulling in? I'm new to the charting changes we're making so thanks for giving some background.

**Comment by lizagilman** [comment]
Hi Dan,
I have two components here - the modal with the chart and the button that opens it
The modal body contains the c3 chart (implemented in `<PieChart> `component)
This replaces the flot/ jquey implementation of fact distribution charts
I added the modal component to storybook
Thanks!

**Review by danseethaler — APPROVED** [review]
This looks great. I tried out the storybook which works well. I am looking forward to see how these changes will be integrated into some pages. Looks great for now!

File: webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        icontype="error-circle-o"
+        key={`message-${factChart.id}`}
+      />
+    );
+
+    return (
+      <Modal show={factChart.showModal} onHide={this.closeModal}>
+        <Modal.Header>
+          <Modal.Title>
+            <b>{__(`Fact distribution chart - ${factChart.title}`)}</b>
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: I think this needs to be extracted differently, probably using sprintf, see javascript section under http://projects.theforeman.org/projects/foreman/wiki/Translating

File: webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        <Modal.Footer>
+          <Button onClick={this.closeModal}>Close</Button>
+        </Modal.Footer>
+      </Modal>
+    );
+  }
+}
+
+const mapStateToProps = (state, ownProps) => ({
+  factChart: ownProps.factChart || state.factChart,
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: when do you get ownProps with factChart vs state?

File: webpack/assets/javascripts/react_app/components/factChart/index.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    getChartData(data.path, data.id);
+    showModal(data.title);
+  }
+
+  closeModal() {
+    this.props.closeModal();
+  }
+
+  render() {
+    return (<Button onClick={this.onClick}>View Chart</Button>);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factChart/index.js]: please extract the string

File: webpack/assets/javascripts/react_app/redux/actions/factCharts/index.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import {
+  FACT_CHART_DATA_REQUEST,
+  FACT_CHART_DATA_SUCCESS,
+  FACT_CHART_DATA_FAILURE,
+  OPEN_MODAL,
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/factCharts/index.js]: Action names are global - OPEN/CLOSE are too generic names, it should be specific to the page / action

File: webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+const initialState = Immutable({
+  showModal: false,
+  chartData: [],
+  loaderStatus: 'PENDING',
+  title: '',
+});
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case FACT_CHART_DATA_SUCCESS:
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js]: where are the implementation for FACT_CHART_DATA_FAILURE and FACT_CHART_DATA_REQUEST?

**Review by ohadlevy — CHANGES_REQUESTED** [review]
thanks @lizagilman i did an initial code review - looking good :-)

added some inline comments, and of course, looking forward to seeing tests :)

File: webpack/assets/javascripts/react_app/components/componentRegistry.js [hunk 6]
```diff
@@ -55,15 +56,17 @@ const componentRegistry = {
   },
 };
 
-const coreComponets = [
+const coreComponents = [
```
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/componentRegistry.js]: Thanks for fixing that typo! 👍

File: webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  render() {
+    const { factChart } = this.props;
+
+    const chart = <PieChart data={factChart.chartData} key={`pie-chart-${factChart.id}`}/>;
+
+    const error = (
+      <MessageBox
+        msg={__('No data available')}
+        icontype="error-circle-o"
```
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: I think we should use icons from [patternfly's icon list](https://www.patternfly.org/styles/icons/).  Perhaps `pficon-error-circle-o`?

File: webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        <Modal.Header>
+          <Modal.Title>
+            <b>{__(`Fact distribution chart - ${factChart.title}`)}</b>
+          </Modal.Title>
+        </Modal.Header>
+        <Modal.Body>
+          <Loader status={factChart.loaderStatus}>{[chart, error]}</Loader>
+        </Modal.Body>
+        <Modal.Footer>
+          <Button onClick={this.closeModal}>Close</Button>
```
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: This is not translated.

File: webpack/assets/javascripts/react_app/components/factChart/index.js (hunk 3 above)
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/factChart/index.js]: This is not translated.

**Review by waldenraines — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js (hunk 2 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: this makes the componenet also resuable by just passing props without dealing with state
initially this was sort of a workaround to get storybook to display the component as we couldn't reach redux state in storybook
@sharvit

File: webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js (hunk 5 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js]: pie chart component displays spinner while waiting and an error message when there's no data.
what should we do in request and failure here?

File: webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js (hunk 7 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: I thinks it's already a pf icon at `<MessageBox>`

Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/factChart/factChartModal/index.js]: Oh I see.  I'm not a fan of how the `<MessageBox>` [restricts you](https://github.com/theforeman/foreman/blob/a5a6b5df465560729c86fa4799c1e9ad13941444/webpack/assets/javascripts/react_app/components/common/MessageBox.js#L11) to using `pficon-${iconType}` (as patternfly also recommends font awesome icons in some cases) but that certainly has nothing to do with this PR.

File: app/views/fact_values/index.html.erb [hunk 9]
```diff
@@ -27,4 +27,7 @@ end
   </tbody>
 </table>
 
+<span id="fact-chart-modal"></span>
+<%= mount_react_component('FactChartModal', "#fact-chart-modal") %>
```
Comment (ohadlevy) [inline app/views/fact_values/index.html.erb]: q: is this really required? e.g. if on Click you would open the modal, does it need a second component mounted already or can you simply open the modal and render your component inside?

File: webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js (hunk 5 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/reducers/factChart/index.js]: IMHO once the action are fired, than we should change the state, for example, on failure - set the correct title etc.
on loading, it should set the loaderStatus correctly etc.

**Comment by waldenraines** [comment]
@lizagilman I wonder if this would be helpful https://github.com/patternfly/patternfly-react/pull/103

File: app/views/fact_values/index.html.erb (hunk 9 above)
Comment (ohadlevy) [reply app/views/fact_values/index.html.erb]: bump?

**Comment by lizagilman** [comment]
updated to use single component (button+modal instead of one for button and one for modal)
added some more stories
some changes in reducer logic 
WIP tests

**Comment by lizagilman** [comment]
add component and reducer tests
WIP action tests

**Comment by lizagilman** [comment]
[test foreman]
[test katello]

**Comment by lizagilman** [comment]
[test foreman]

**Comment by lizagilman** [comment]
[test foreman]

**Comment by lizagilman** [comment]
[test katello]

**Comment by lizagilman** [comment]
@danseethaler / @ohadlevy can you take another look?

**Comment by mmoll** [comment]
[test katello]
[test upgrade]

**Comment by waldenraines** [comment]
[test katello]

File: .storybook/preview-head.html [hunk 10]
```diff
@@ -11,3 +11,12 @@
     return str;
   }
 </script>
+<!-- workaround to get Jed into storybook -->
+<script>
+  window.Jed = {
+    sprintf: function(str, elem) {
```
Comment (amirfefer) [inline .storybook/preview-head.html]: nitpick: use es6 arrow function

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+          :
+          false
+        }
+      </div>
+    );
+  }
+}
+
+const mapStateToProps = (state, ownProps) => ({
+  factChart: ownProps.factChart || state.factChart,
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: You can map the specific dispalyModal state to props (in `mapsStateToProps` function), something like this:
```javascript
displayModal: state.factChart.displayModal[ownProps.id] || false,
```
after this change,  line 32 becomes redundant

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  closeModal() {
+    this.props.closeModal();
+  }
+
+  render() {
+    const { factChart, data: { id, title } } = this.props;
+
+    const showModal = factChart.displayModal[id];
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: its the same name as the action function, shouldn't it has a unique name?
or maybe map `displayModal` to a prop?

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+    const chartConfig = showModal ? getLargePieChartConfig({ data: factChart.chartData }) : false;
+
+    const chart = showModal ? (<Chart
+        className="c3-statistics-pie"
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: Why not using `<PieChart>` component?

File: webpack/assets/javascripts/services/ChartService.consts.js [hunk 14]
```diff
@@ -20,7 +20,7 @@ export const donutChartConfig = {
   tooltip: {
     show: true,
   },
-  legend: { show: false },
+  legend: { show: true },
```
Comment (amirfefer) [inline webpack/assets/javascripts/services/ChartService.consts.js]: shouldn't we add legend on large charts only ?

**Review by amirfefer — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/factCharts/index.js (hunk 13 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factCharts/index.js]: `<PieChart>` is `small` by default

File: webpack/assets/javascripts/services/ChartService.consts.js (hunk 14 above)
Comment (lizagilman) [reply webpack/assets/javascripts/services/ChartService.consts.js]: looks ok to me. how does it look to you?

Comment (amirfefer) [reply webpack/assets/javascripts/services/ChartService.consts.js]: Its impact small size charts as well

In dashboard  you have legend information twice:
donut_chart

IMHO adding legend into a small size chart is a bit redundant, because its too small. instead we could click to enlarge it, and see it.

statistics

File: webpack/assets/javascripts/react_app/components/factCharts/index.js (hunk 13 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/factCharts/index.js]: why not pass it as a prop?

**Comment by lizagilman** [comment]
@amirfefer addressed all your comments

File: .storybook/preview-head.html [hunk 15]
```diff
@@ -11,3 +11,12 @@
     return str;
   }
 </script>
+<!-- workaround to get Jed into storybook -->
+
+<script>
+  const sprintfStr = (str, elem) => { return str } ;
+
+  window.Jed = {
+    sprintf:   sprintfStr
```
Comment (amirfefer) [inline .storybook/preview-head.html]: Does it work well when running stroybook? shouldn't it be a function?

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            </Modal.Header>
+            <Modal.Body>
+              <Loader status={factChart.loaderStatus}>{[chart, error]}</Loader>
+            </Modal.Body>
+            <Modal.Footer>
+              <Button onClick={this.closeModal}>{__('Close')}</Button>
+            </Modal.Footer>
+          </Modal>
+          :
+          false
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: Instead of if statement, you could use 
```javascript
modalToDisplay && <Component>
```

File: webpack/assets/javascripts/react_app/redux/reducers/index.js [hunk 17]
```diff
@@ -5,6 +5,8 @@ import statistics from './statistics';
 import hosts from './hosts';
 import notifications from './notifications/';
 import toasts from './toasts';
+import factChart from './factCharts/';
+
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/reducers/index.js]: redundant line

**Review by amirfefer — CHANGES_REQUESTED** [review]

File: .storybook/preview-head.html (hunk 15 above)
Comment (lizagilman) [reply .storybook/preview-head.html]: Works well, it is sort of a workaround to get Jed work in storybook

**Comment by lizagilman** [comment]
@amirfefer updated

**Comment by lizagilman** [comment]
[test katello]

**Comment by lizagilman** [comment]
@danseethaler any comments?

**Comment by lizagilman** [comment]
@amirfefer ping

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.fixtures.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+export const initialState = Immutable({
+  factChart: {
+    modalToDisplay: {},
+    chartData: mockStoryData.config.data.columns,
+    loaderStatus: '',
+  },
+});
+
+export const modalOpenState = Immutable({
+  factChart: {
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/factChart.fixtures.js]: Similar repetitive blocks, could you dry it a bit?

File: .storybook/preview-head.html (hunk 15 above)
Comment (amirfefer) [inline .storybook/preview-head.html]: nitpick: redundant space?

File: app/controllers/facts_controller.rb [hunk 19]
```diff
@@ -13,6 +13,7 @@ def show
     rescue
       data = @fact
     end
+    data[:values] = data[:values].map { |x| x.values }
```
Comment (amirfefer) [inline app/controllers/facts_controller.rb]: what about something like this: `data[:values].map!(&:values)`?
`map!` replaces the element

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+import { STATUS } from '../../constants';
+import * as FactChartActions from '../../redux/actions/factCharts/';
+
+class FactChart extends React.Component {
+  constructor(props) {
+    super(props);
+    helpers.bindMethods(this, ['onClick', 'closeModal']);
+  }
+
+  onClick() {
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: could you pick a more informative name?

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+  render() {
+    const { factChart, data: { id, title }, modalToDisplay } = this.props;
+
+    const chart = modalToDisplay ? (<PieChart
+        data={factChart.chartData}
+        key={`chart-${id}`}
+        smallChart = {false}
+      />) : false;
+
+    const requestErrorMsg = factChart.loaderStatus === STATUS.ERROR ? __('Request Failed') : undefined;
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: what do you think about moving `__('No data available')` to here instead of undefined?

File: webpack/assets/javascripts/react_app/redux/reducers/factCharts/factChart.fixtures.js [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+export const chartDataValues = [['Fedora 21', 3], ['Ubuntu 14.04', 4], ['Centos 7', 2], ['Debian 8', 1]];
+
+export const initialState = Immutable({
+  modalToDisplay: {},
+  chartData: [],
+  loaderStatus: '',
+});
+
+export const afterModalOpen = Immutable({
+  modalToDisplay: { 1: true },
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/reducers/factCharts/factChart.fixtures.js]: could you dry it a bit as well? some blocks are repeated here.

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js [hunk 23]
```diff
@@ -0,0 +1,19 @@
+import { mount, shallow } from 'enzyme';
+import React from 'react';
+import FactChart from './';
+import Store from '../../redux';
+
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js]: could you use snapshot testing here?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @lizagilman ! 
almost there !
I left some inline comments.

**Comment by lizagilman** [comment]
[test foreman]

**Comment by lizagilman** [comment]
updated

**Review by amirfefer — CHANGES_REQUESTED** [review]
LGTM! 
one last thing - could you please rename the commit title? 
it's not describe this PR.
something like : "Use pf-react donut chart in facts page" or so.
btw tests are red, try rebase?

**Comment by lizagilman** [comment]
@amirfefer thanks, renamed
this branch is up to date, I think this is not related.

**Comment by amirfefer** [comment]
@lizagilman the commit message still old

**Review by amirfefer — APPROVED** [review]
Thanks @lizagilman 
LGTM

File: webpack/assets/javascripts/react_app/components/componentRegistry.js [hunk 24]
```diff
@@ -8,6 +8,8 @@ import ToastsList from './toastNotifications/';
 import StorageContainer from './hosts/storage/vmware/';
 import BookmarkContainer from './bookmarks';
 import PasswordStrength from './user/passwordStrength/';
+// eslint-disable-next-line
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/componentRegistry.js]: is this required?

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js [hunk 25]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { storiesOf } from '@storybook/react';
+import thunk from 'redux-thunk';
+import configureMockStore from 'redux-mock-store';
+// eslint-disable-next-line
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js]: also not sure if its required?

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+import Immutable from 'seamless-immutable';
+import PropTypes from 'prop-types';
+import helpers from '../../common/helpers';
+import DonutChart from '../common/charts/DonutChart';
+import Loader from '../common/Loader';
+import MessageBox from '../common/MessageBox';
+import { STATUS } from '../../constants';
+import * as FactChartActions from '../../redux/actions/factCharts/';
+
+
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: extra whitespaces.

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+    const requestErrorMsg = factChart.loaderStatus === STATUS.ERROR ? __('Request Failed') : __('No data available');
+
+    const error = modalToDisplay ? (
+        <MessageBox
+          msg={ requestErrorMsg }
+          icontype="error-circle-o"
+          key={`message-${id}`}
+        />) : false;
+
+
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: extra whitespaces - can you make sure you run those changes with prettier ?

**Review by ohadlevy — CHANGES_REQUESTED** [review]
Thanks @lizagilman I've left a few comments inline, beside those I'm missing a few things:

1. clicking on the chart does not redirect to search.
2. I'm missing some UI (e.g. header, close "X" btn etc) - see the current screenshot below:

image

**Comment by lizagilman** [comment]
@ohadlevy updated

File: webpack/assets/javascripts/react_app/components/componentRegistry.js (hunk 24 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/componentRegistry.js]: yes, otherwise there's an error - Using exported name 'FactChart' as identifier for default export
but I do want to use that name

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js (hunk 25 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js]: same here

**Comment by ohadlevy** [comment]
@lizagilman please make sure to update the snapshots too once they are changed.

File: .storybook/preview-head.html [hunk 28]
```diff
@@ -11,3 +11,12 @@
     return str;
   }
 </script>
+<!-- workaround to get Jed into storybook -->
+
+<script>
+  const sprintfStr = (str, elem) => { return str } ;
```
Comment (ohadlevy) [inline .storybook/preview-head.html]: nitpick: this is a static html / js script tag which does not go though babel, maybe we should not use ES6 here as it would limit storybook to recent browsers.

**Comment by lizagilman** [comment]
@ohadlevy just noticed
updated

File: webpack/assets/javascripts/react_app/components/componentRegistry.js (hunk 24 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/componentRegistry.js]: i think this happens because you export twice in factCharts, i believe you only need to export the default one.

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js (hunk 25 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/factCharts/factChart.stories.js]: same here too :)

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js [hunk 29]
```diff
@@ -0,0 +1,30 @@
+import { shallow } from 'enzyme';
+import toJson from 'enzyme-to-json';
+import React from 'react';
+import ConnectedFactChart, { FactChart } from './';
+import Store from '../../redux';
+
+const state = {
+  chartData: [],
+  loaderStatus: '',
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js]: why empty string? shouldn't this be a const of sort?

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+    expect(toJson(wrapper)).toMatchSnapshot();
+  });
+
+  it('should open modal after click', () => {
+    const props = { onClick: jest.fn(), data: { id: 1 } };
+    const wrapper = shallow(<ConnectedFactChart store={ Store } { ...props } />);
+
+    wrapper.simulate('click');
+    expect(props.onClick).toBeCalled();
+  });
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js]: missing tests for searching?

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+      <MessageBox
+        msg={requestErrorMsg}
+        icontype='error-circle-o'
+        key={`message-${id}`}
+      />
+    ) : (
+      false
+    );
+
+    const tooltip = factChart.chartData ? {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: imho you should use react tooltip, not jquery

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+      title: __('Show distribution chart'),
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js]: missing the host counter (see the screenshot i attached originally)

**Review by ohadlevy — CHANGES_REQUESTED** [review]
please also delete the old view files / js as part of this pr.

**Comment by ohadlevy** [comment]
@amirfefer please review again - thanks.

**Comment by lizagilman** [comment]
@amirfefer @ohadlevy updated
@amirfefer can you confirm it is safe to delete `app/assets/javascripts/charts.js `?

File: webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js (hunk 30 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factCharts/factChart.test.js]: shouldn't this functionality be tested at `ChartService`?

File: webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+});
+
+let hostsCount = 0;
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case FACT_CHART_DATA_REQUEST:
+      return state.set('loaderStatus', 'PENDING');
+    case FACT_CHART_DATA_SUCCESS:
+      action.payload.values.forEach((val) => { (hostsCount += val[1]); });
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js:24]: this is probably a good candidate to introduce https://github.com/reactjs/reselect - @sharvit @amirfefer what do you guys think?

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+    const handleChartClick =
+      search && search.match(/=$/) ? null : navigateToSearch.bind(null, search);
+
+    const chartProps = {
+      data: factChart.chartData ? factChart.chartData : null,
+      key: `chart-${id}`,
+      onclick: handleChartClick,
+    };
+
+    const chart = <DonutChart {...chartProps} config='large' />;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js:47]: if the field has a long name, it's not readable, I'm not sure if its the c3 impementation (e.g. we need to open an issue on patternfly-react or somewhere else) but this its how it looks:
image

vs

image

File: webpack/assets/javascripts/react_app/components/factCharts/index.js [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+          <Modal show={true} onHide={this.closeModal}>
+            <Modal.Header closeButton>
+              <Modal.Title>
+                <b>
+                  {// eslint-disable-next-line no-undef
+                  Jed.sprintf(__('Fact distribution chart - %s '), title)}
+                </b>
+                <small>
+                  {// eslint-disable-next-line no-undef
+                  Jed.sprintf(
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/factCharts/index.js:83]: please add parenthesis around the counter (e.g. like in this screenshot) 
image

**Comment by ohadlevy** [comment]
@lizagilman regarding removing charts.js / flot - please leave it and instead please fix http://projects.theforeman.org/issues/23010 - thanks :)

File: webpack/assets/javascripts/react_app/components/factCharts/index.js (hunk 35 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factCharts/index.js:83]: added

File: webpack/assets/javascripts/react_app/components/factCharts/index.js (hunk 34 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/factCharts/index.js:47]: I and @amirfefer think we should open an issue on that

**Comment by ohadlevy** [comment]
can we limit the text using ellipsis or similar until its resolved?

**Comment by amirfefer** [comment]
@ohadlevy  as far as i know, we can pass trim values, but the tooltip won't show with the full value.
**EDIT**: after a quick look we can pass [tooltip.name](http://c3js.org/reference.html#tooltip-format-name-props) with the full value, and trim the values.

**Comment by lizagilman** [comment]
@ohadlevy @amirfefer updated,
we now have the shortened labels and tooltips with full text 

legend_tooltip

screenshot from 2018-03-28 16-00-27

File: webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js (hunk 33 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js:24]: @sharvit @amirfefer @waldenraines @tstrachota  ping?

**Comment by lizagilman** [comment]
@ohadlevy @amirfefer ping?

**Review by ohadlevy — DISMISSED** [review]
thanks, this looks good, I've rebased and pushed so tests can run again, once they pass, I'll merge.

**Comment by ohadlevy** [comment]
@amirfefer @sharvit any idea on why tests might be failing?

File: webpack/assets/javascripts/services/ChartService.js [hunk 36]
```diff
@@ -30,15 +30,48 @@ const getChartConfig = ({
...
+  const dataExists = doDataExist(data);
+
+  const longNames = [];
+  const shortNames = [];
+
+  let dataWithShortNames = [];
+
+  if (dataExists) {
+    dataWithShortNames = data.map((val) => {
+      const item = val.slice().asMutable();
```
Comment (ohadlevy) [inline webpack/assets/javascripts/services/ChartService.js]: note this breaks the dashboard page.

**Comment by amirfefer** [comment]
@ohadlevy ,
Note that  there is a js error in the dashboard page, which caught by `dashboard_integration_test` .
Looks like this [change](https://github.com/theforeman/foreman/pull/5019/files#diff-1301a72a99b657c8c32ace90d1d5c670R42) causes this issue.

**Comment by lizagilman** [comment]
[test foreman]
[test katello]

**Comment by lizagilman** [comment]
[test foreman]

**Comment by lizagilman** [comment]
[test foreman]

**Comment by lizagilman** [comment]
@amirfefer @ohadlevy tests are green

**Comment by lizagilman** [comment]
@amirfefer any other comments?

**Comment by amirfefer** [comment]
there are some js warnings when a modal opens:
screenshot from 2018-05-03 15-57-01

I fixed this issue [here](https://github.com/theforeman/foreman/pull/5524)

**Comment by amirfefer** [comment]
The modal opens with a full black background: 
with-backdrop

but when we pass `{backdrop={false}` the background is visible:
with-no-backdrop

@Rohoover, should the background be still visible?

**Review by Rohoover — COMMENTED** [review]
@amirfefer 

It should neither be black or clear. 

See the PF storybook modal for the correct transparency for modals:

https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?selectedKind=About%20Modal&selectedStory=AboutModal&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

Sorry I don't have the transparency percentage off the top of my head.

**Comment by lizagilman** [comment]
rebased, 
no warnings, and modal looks normal - half transparent (no black background)

**Review by amirfefer — APPROVED** [review]
Thanks! LGFM :+1:

**Comment by lizagilman** [comment]
@tstrachota can we merge? any more comments?

File: webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js (hunk 33 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/reducers/factCharts/index.js:24]: @lizagilman can you please create a follow up issue on redmine so we don't miss this?

File: webpack/assets/javascripts/services/ChartService.js [hunk 37]
```diff
@@ -30,15 +31,48 @@ const getChartConfig = ({
...
   return {
     ...chartConfigForType,
     id,
     data: {
-      columns: doDataExist(data) ? data : [],
+      columns: dataExists ? dataWithShortNames : [],
       onclick,
       ...(colorsSize > 0 ? { colors } : {}),
     },
+    // eslint-disable-next-line
```
Comment (ohadlevy) [inline webpack/assets/javascripts/services/ChartService.js]: why did you need to disable all linting here? you can probably just use 
```js
    // eslint-disable-next-line no-shadow
````

**Review by ohadlevy — COMMENTED** [review]
This looks fairly good, I think I have one last request, when opening / closing a modal, we always make an API call, maybe we can make the api call only if we don't have the data?

**Comment by lizagilman** [comment]
I think this issue has come up already, someone mentioned that if we cache the data here users won't be able to see the most recent data, and they would want to get the updated data each time they open the modal. One thing I can think of is maybe we can add a refresh button inside the modal.

File: webpack/assets/javascripts/services/ChartService.js (hunk 37 above)
Comment (lizagilman) [reply webpack/assets/javascripts/services/ChartService.js]: updated

**Comment by ohadlevy** [comment]
Ok, let's keep as is, in either cases I don't expect someone to stay long
enough in the index page that it would matter that much.

**Review by ohadlevy — APPROVED** [review]
